File: Bookcase core/src/main/java/cz/vhromada/book/validator/AuthorValidator.java
package cz.vhromada.book.validator;

import cz.vhromada.book.entity.Author;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.validator.AbstractMovableValidator;
import cz.vhromada.result.Event;
import cz.vhromada.result.Result;
import cz.vhromada.result.Severity;

import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;

/**
 * A class represents implementation of validator for author.
 *
 * @author Vladimir Hromada
 */
@Component("authorValidator")
public class AuthorValidator extends AbstractMovableValidator<Author, cz.vhromada.book.domain.Author> {

    /**
     * Creates a new instance of AuthorValidator.
     *
     * @param authorService service for authors
     * @throws IllegalArgumentException if service for authors is null
     */
    public AuthorValidator(final MovableService<cz.vhromada.book.domain.Author> authorService) {
        super("Author", authorService);
    }

    /**
     * Validates author deeply.
     * <br>
     * Validation errors:
     * <ul>
     * <li>First name is null</li>
     * <li>First name is empty string</li>
     * <li>Middle name is empty string</li>
     * <li>Last name is null</li>
     * <li>Last name is empty string</li>
     * </ul>
     *
     * @param data   validating game
     * @param result result with validation errors
     */
    @Override
    protected void validateDataDeep(final Author data, final Result<Void> result) {
        if (data.getFirstName() == null) {
            result.addEvent(new Event(Severity.ERROR, "AUTHOR_FIRST_NAME_NULL", "First name mustn't be null."));
        } else if (!StringUtils.hasText(data.getFirstName())) {
            result.addEvent(new Event(Severity.ERROR, "AUTHOR_FIRST_NAME_EMPTY", "First name mustn't be empty string."));
        }
        if (data.getMiddleName() != null && !StringUtils.hasText(data.getMiddleName().trim())) {
            result.addEvent(new Event(Severity.ERROR, "AUTHOR_MIDDLE_NAME_EMPTY", "Middle name mustn't be empty string."));
        }
        if (data.getLastName() == null) {
            result.addEvent(new Event(Severity.ERROR, "AUTHOR_LAST_NAME_NULL", "Last name mustn't be null."));
        } else if (!StringUtils.hasText(data.getLastName())) {
            result.addEvent(new Event(Severity.ERROR, "AUTHOR_LAST_NAME_EMPTY", "Last name mustn't be empty string."));
        }
    }

}

File: bookcase-core/src/main/kotlin/cz/vhromada/bookcase/validator/AuthorValidator.kt
package cz.vhromada.bookcase.validator

import cz.vhromada.bookcase.entity.Author
import cz.vhromada.common.result.Event
import cz.vhromada.common.result.Result
import cz.vhromada.common.result.Severity
import cz.vhromada.common.service.MovableService
import cz.vhromada.common.validator.AbstractMovableValidator
import org.springframework.stereotype.Component

/**
 * A class represents validator for author.
 *
 * @author Vladimir Hromada
 */
@Component("authorValidator")
class AuthorValidator(authorService: MovableService<cz.vhromada.bookcase.domain.Author>) : AbstractMovableValidator<Author, cz.vhromada.bookcase.domain.Author>("Author", authorService) {

    /**
     * Validates author deeply.
     * <br></br>
     * Validation errors:
     *
     *  * First name is null
     *  * First name is empty string
     *  * Middle name is null
     *  * Last name is null
     *  * Last name is empty string
     *
     * @param data   validating author
     * @param result result with validation errors
     */
    override fun validateDataDeep(data: Author, result: Result<Unit>) {
        when {
            data.firstName == null -> {
                result.addEvent(Event(Severity.ERROR, "AUTHOR_FIRST_NAME_NULL", "First name mustn't be null."))
            }
            data.firstName.isBlank() -> {
                result.addEvent(Event(Severity.ERROR, "AUTHOR_FIRST_NAME_EMPTY", "First name mustn't be empty string."))
            }
        }
        if (data.middleName == null) {
            result.addEvent(Event(Severity.ERROR, "AUTHOR_MIDDLE_NAME_NULL", "Middle name mustn't be null."))
        }
        when {
            data.lastName == null -> {
                result.addEvent(Event(Severity.ERROR, "AUTHOR_LAST_NAME_NULL", "Last name mustn't be null."))
            }
            data.lastName.isBlank() -> {
                result.addEvent(Event(Severity.ERROR, "AUTHOR_LAST_NAME_EMPTY", "Last name mustn't be empty string."))
            }
        }
    }

}
